pallets/dca/src/lib.rs

```diff
@@ -0,0 +1,995 @@
+
+			Ok(*current_id)
+		})
+	}
+
+	fn get_next_block_number() -> BlockNumberFor<T> {
+		let mut current_block_number = frame_system::Pallet::<T>::current_block_number();
+		current_block_number.saturating_inc();
+
+		current_block_number
+	}
+
+	fn create_bounded_vec(
+		next_schedule_id: ScheduleId,
+	) -> Result<BoundedVec<ScheduleId, T::MaxSchedulePerBlock>, DispatchError> {
+		let schedule_id = vec![next_schedule_id];
+		let bounded_vec: BoundedVec<ScheduleId, T::MaxSchedulePerBlock> =
+			schedule_id.try_into().map_err(|_| Error::<T>::InvalidState)?;
+		Ok(bounded_vec)
+	}
+
+	fn execute_trade(origin: T::Origin, order: &Order<<T as Config>::Asset>) -> DispatchResult {
+		match order {
+			Order::Sell {
+				asset_in,
+				asset_out,
+				amount_in,
+				min_limit,
+				route: _,
+			} => {
+				let min_limit_with_slippage = Self::get_min_limit_with_slippage(asset_in, asset_out, amount_in)?;
+
+				let transaction_fee = Self::get_transaction_fee(*asset_in)?;
+
+				let amount_to_sell = amount_in
+					.checked_sub(transaction_fee)
+					.ok_or(ArithmeticError::Underflow)?;
+
+				T::AMMTrader::sell(
+					origin,
+					*asset_in,
+					*asset_out,
+					amount_to_sell,
+					min(*min_limit, min_limit_with_slippage),
+				)
+			}
+			Order::Buy {
+				asset_in,
+				asset_out,
+				amount_out,
+				max_limit,
+				route: _,
+			} => {
+				let max_limit_with_slippage = Self::get_max_limit_with_slippage(asset_in, asset_out, amount_out)?;
+
+				T::AMMTrader::buy(
+					origin,
+					*asset_in,
+					*asset_out,
+					*amount_out,
+					max(*max_limit, max_limit_with_slippage),
+				)
+			}
+		}
+	}
+
+	fn get_min_limit_with_slippage(
+		asset_in: &<T as Config>::Asset,
+		asset_out: &<T as Config>::Asset,
+		amount_in: &Balance,
+	) -> Result<u128, DispatchError> {
+		let price = Self::get_price_from_last_block_oracle(*asset_in, *asset_out)?;
+
+		let estimated_amount_out = price.checked_mul_int(*amount_in).ok_or(ArithmeticError::Overflow)?;
+
+		let slippage_amount = T::SlippageLimitPercentage::get().mul_floor(estimated_amount_out);
+		let min_limit_with_slippage = estimated_amount_out
+			.checked_sub(slippage_amount)
+			.ok_or(ArithmeticError::Overflow)?;
+
+		Ok(min_limit_with_slippage)
+	}
+
+	fn get_max_limit_with_slippage(
+		asset_in: &<T as Config>::Asset,
+		asset_out: &<T as Config>::Asset,
+		amount_out: &Balance,
+	) -> Result<u128, DispatchError> {
+		let price = Self::get_price_from_last_block_oracle(*asset_out, *asset_in)?;
+
+		let estimated_amount_in = price.checked_mul_int(*amount_out).ok_or(ArithmeticError::Overflow)?;
+
+		let slippage_amount = T::SlippageLimitPercentage::get().mul_floor(estimated_amount_in);
+		let max_limit_with_slippage = estimated_amount_in
+			.checked_add(slippage_amount)
+			.ok_or(ArithmeticError::Overflow)?;
+
+		Ok(max_limit_with_slippage)
+	}
+
+	fn get_transaction_fee(fee_currency: T::Asset) -> Result<u128, DispatchError> {
```

Comment:
just chargin user for on_initialize weight is not sufficient, they have to be charged based on their trade as, eg. it should be more expensive than swap tx itself for example